External products: external link & button text

jaclync opened the pull request:
"editable product link" for #2200 

## Changes

- Encoded `externalURL` and `buttonText` for updating an external product
- Added a function to update `externalURL` and `buttonText` in `ProductFormViewModel` + test case
- Created `ProductExternalLinkViewController` that enabled editing the external URL and button text fields for an external product
- Created an enum `TextFieldContentAlignment` for different ways of text field content alignment for the text and placeholder (I was debating between this name and `TextFieldTextAlignment`, lemme know if you think another name would sound more clear!)
  - Because `NSTextAlignment` only has `left`/`right` instead of `leading`/`trailing` that takes LTR/RTL layout into account, this enum is mapped to `NSTextAlignment` based on the app layout
- Added `keyboardType` and `textFieldAlignment` parameters to `TitleAndTextFieldTableViewCell` for customizations in the external link screen
- Enabled editing external link in `ProductFormViewController`

## Testing

- Go to the Products tab
- Tap on an external product
- Tap on the Product link row --> should see the screen with two text fields for external URL and button text as in the screenshots
- Edit the external URL and button text
- (If not on iOS 13.4 simulator/device) Tap "<" in the navigation bar --> a discard changes prompt should be shown
- Tap "Cancel" to keep editing
- Tap "Done" to save changes
- Tap "Update" --> the external URL and button text should be updated remotely

## Example screenshots

![simulator](https://user-images.githubusercontent.com/1945542/84986133-973c8200-b170-11ea-8162-27d3d0751681.gif)


empty state | filled state
-- | --
![Simulator Screen Shot - iPhone 11 - 2020-06-18 at 10 34 25](https://user-images.githubusercontent.com/1945542/84972500-8e888380-b151-11ea-84b0-b2e687b50d1c.png) | ![Simulator Screen Shot - iPhone 11 - 2020-06-18 at 10 34 48](https://user-images.githubusercontent.com/1945542/84972506-91837400-b151-11ea-9700-ae38a2317e04.png)
![Simulator Screen Shot - iPhone 11 - 2020-06-18 at 10 35 33](https://user-images.githubusercontent.com/1945542/84972509-934d3780-b151-11ea-820f-e92116cdec95.png) | ![Simulator Screen Shot - iPhone 11 - 2020-06-18 at 10 35 28](https://user-images.githubusercontent.com/1945542/84972508-92b4a100-b151-11ea-8428-97a1c37b9b4d.png)


![Simulator Screen Shot - iPhone 11 - 2020-06-18 at 10 33 48](https://user-images.githubusercontent.com/1945542/84972519-97795500-b151-11ea-98a1-c95dc28a968d.png)


Update release notes:
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

jaclync commented:
> if users add the button url but not the button text, is the product visible? I just remembe that's why initially I had a default "Buy now" button text already filled out instead of the "Enter button text" placeholder, can we please go back to this behavior?

@Garance91540 I just tested the behavior in core, "Buy product" is shown on the shop page (at least for my site) when the button text field is empty with "Buy product" as its placeholder. I think it'd be nice to follow core, to update the button text field placeholder to "Buy product"? Automatically filling in the "Buy now" or "Buy product" button text feels a bit surprising if I were the user 🤔  Also, the site could have a different localization setting from the device.

<img width="662" alt="Screen Shot 2020-06-18 at 4 29 12 PM" src="https://user-images.githubusercontent.com/1945542/84997989-d32c1300-b181-11ea-817c-49e256804b47.png">

> Do we display an error message under the "add link" text field if users enter an invalid url?

Good point! we could add some local validation for the URL, but it wouldn't cover all possible malformed URLs or URLs that have no response or any network errors. Still, it is better than allowing any text. Do we have design for the error state? (maybe like the duplicate SKU in the inventory settings)

shiki commented on WooCommerce/Classes/ViewRelated/ReusableViews/TextFieldContentAlignment.swift:
```diff
@@ -0,0 +1,22 @@
+import UIKit
+
+/// Horizontal alignment for content (text or placeholder) in a text field.
+enum TextFieldContentAlignment {
```
⚠️  As you implied, the name is a bit confusing. 😬 But I think what would be more valuable is to explain the use cases of this `enum`. Why and when would we want to use this instead of just `NSTextAlignment`? What do you think about adding a comment about that?

jaclync replied:
I renamed this to `TextFieldTextAlignment` and updated the comment in https://github.com/woocommerce/woocommerce-ios/pull/2442/commits/1e378d324ae0088c28309c6471d46615905e6988 - lemme know if any part can still be improved!

shiki replied:
Awesome! The description and the unit tests really help!

shiki commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit External Link/ProductExternalLinkViewController.swift:
```diff
@@ -0,0 +1,232 @@
+private extension ProductExternalLinkViewController {
+    /// Registers for all of the related Notifications
+    ///
+    func startListeningToNotifications() {
```
I would not be opposed to using a longer but more specific name so when reading this inside `viewDidLoad`, it's much clearer what to expect. 

```suggestion
    func startObservingKeyboardNotifications() {
```

I'd also argue that it's probably not worth it to create a method for this since it's just one line. It probably makes the class easier to read since it's shorter? 🤷 Up to you though.

jaclync replied:
I think it's still nice having a descriptive function even though the code is just one line. I like the name you suggested, updated in https://github.com/woocommerce/woocommerce-ios/pull/2442/commits/8a9971de59fadf9b837a9e21d92a724e23627d93

shiki commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit External Link/ProductExternalLinkViewController.swift:
```diff
@@ -0,0 +1,232 @@
+    private var externalURL: String?
+    private var buttonText: String
+
+    private lazy var keyboardFrameObserver: KeyboardFrameObserver = {
+        let keyboardFrameObserver = KeyboardFrameObserver { [weak self] keyboardFrame in
+            self?.handleKeyboardFrameUpdate(keyboardFrame: keyboardFrame)
+        }
+        return keyboardFrameObserver
+    }()
```
You could make this shorter by:

```suggestion
    private lazy var keyboardFrameObserver = KeyboardFrameObserver { [weak self] keyboardFrame in
        self?.handleKeyboardFrameUpdate(keyboardFrame: keyboardFrame)
    }

```

jaclync replied:
nice! updated in https://github.com/woocommerce/woocommerce-ios/pull/2442/commits/8a9971de59fadf9b837a9e21d92a724e23627d93

shiki commented on WooCommerce/Classes/ViewRelated/ReusableViews/TextFieldContentAlignment.swift:
```diff
@@ -0,0 +1,22 @@
+}
+
+extension TextFieldContentAlignment {
+    func toTextAlignment(layoutDirection: UIUserInterfaceLayoutDirection = UIApplication.shared.userInterfaceLayoutDirection) -> NSTextAlignment {
```
Were you planning to add a test for this? Is that why you exposed the `layoutDirection` argument?

jaclync replied:
yes I did plan to add tests for this, but thought this mapping feels straightforward. it's easy to add tests though, so I added them in https://github.com/woocommerce/woocommerce-ios/pull/2442/commits/1e378d324ae0088c28309c6471d46615905e6988

shiki commented on WooCommerce/WooCommerceTests/ViewRelated/Products/Edit Product/ProductFormViewModel+UpdatesTests.swift:
```diff
@@ -225,6 +225,25 @@ final class ProductFormViewModel_UpdatesTests: XCTestCase {
+
+    func testUpdatingExternalLink() {
+        // Arrange
+        let product = MockProduct().product(sku: "")
```
You probably just wanted: 

```suggestion
        let product = MockProduct().product()
```

😁

jaclync replied:
cool! updated in https://github.com/woocommerce/woocommerce-ios/pull/2442/commits/bc3de89a077dd574de02c7cb19a2e9c71e17b368

shiki commented:
By the way, good separation of commits in this PR. 👍

jaclync commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit External Link/ProductExternalLinkViewController.swift:
```diff
@@ -0,0 +1,208 @@
+
+    func configureExternalURL(cell: TitleAndTextFieldTableViewCell) {
+        let title = NSLocalizedString("Product URL", comment: "")
+        let placeholder = NSLocalizedString("Enter URL", comment: "")
```
oops I tend to test out the UI and forgot to add a comment 🤦🏻‍♀️  thanks for catching this! updated in https://github.com/woocommerce/woocommerce-ios/pull/2442/commits/35c60bafe9089c6d403129dfff1e56ac490dd72b

jaclync commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit External Link/ProductExternalLinkViewController.swift:
```diff
@@ -0,0 +1,208 @@
+
+    func configureButtonText(cell: TitleAndTextFieldTableViewCell) {
+        let title = NSLocalizedString("Button Text", comment: "")
+        let placeholder = NSLocalizedString("Enter button text", comment: "")
```
haha yea, thanks for catching this! updated in https://github.com/woocommerce/woocommerce-ios/pull/2442/commits/35c60bafe9089c6d403129dfff1e56ac490dd72b

jaclync commented:
@shiki thanks for the review! I updated the PR and responded to all your comments for another look 😄 

> Are you planning to work on the URL validation on another PR? That sounds better to make this simpler

yes, I added a subtask for URL validation to #2200 description and will handle this in a separate PR.

@Garance91540 I updated the button text placeholder following core:

<img src="https://user-images.githubusercontent.com/1945542/85087992-91e34400-b211-11ea-8e04-7990170b02b1.png" width="320" />

Garance91540 commented:
> @Garance91540 I updated the button text placeholder following core:

@jaclync Great! can we make the "Buy Product" color iOS System Label/ Primary Color? (right now it seems like it's disabled)

jaclync commented:
> @jaclync Great! can we make the "Buy Product" color iOS System Label/ Primary Color? (right now it seems like it's disabled)

@Garance91540 it's the disabled color since I set "Buy product" as the placeholder for the button text field, just like in core. did you mean we want to fill out the button text if it's currently empty, which is then saved to the product if the user taps "Done" on this screen? If so, I have some concerns about this:

- If a product doesn't have the button text field set before coming to this screen, and we automatically set this field, then it has unsaved changes immediately. since we show the discard changes prompt when they try exiting the screen without saving, it could be confusing if the user hasn't manually changed anything
- I felt leaving the default button text to core (the default text that is shown on the shop page if the button text field isn't set) would be more ideal. If our assumption of the default button text isn't the same as in core for certain languages, then it could be confusing if they have different default button text for various external products

jaclync commented:
@Garance91540 since the PR has been reviewed and tested, also with WWDC happening, I'm leaving the remaining followup tasks from this PR to 2 separate subtasks for #2200 🙂 

the 2 remaining subtasks (also in the issue description):
- editing product link followup: local validation for external URL
- editing product link followup: auto-fill button text with "Buy product" when the external URL field becomes non-empty
